constraint-solver/src/solver.rs

```diff
@@ -62,23 +73,37 @@ impl<T: FieldElement, V: Ord + Clone + Hash + Eq + Display + Debug> Solver<T, V>
             .collect();
         Ok(SolveResult {
             assignments,
-            simplified_algebraic_constraints: self.algebraic_constraints,
+            simplified_constraint_system: self.constraint_system,
         })
     }
 
     fn loop_until_no_progress(&mut self) -> Result<(), Error> {
         loop {
             let mut progress = false;
-            for i in 0..self.algebraic_constraints.len() {
+            for i in 0..self.constraint_system.algebraic_constraints.len() {
                 // TODO: Improve efficiency by only running skipping constraints that
                 // have not received any updates since they were last processed.
-                let effects = self.algebraic_constraints[i]
+                let effects = self.constraint_system.algebraic_constraints[i]
                     .solve(&self.range_constraints)?
                     .effects;
                 for effect in effects {
                     progress |= self.apply_effect(effect);
                 }
             }
+            if let Some(bus_interaction_handler) = &self.bus_interaction_handler {
+                let effects = self
+                    .constraint_system
+                    .bus_interactions
+                    .iter()
+                    .flat_map(|bus_interaction| {
+                        bus_interaction.solve(&**bus_interaction_handler, &self.range_constraints)
+                    })
+                    // Collect to satisfy borrow checker
```

Comment:
I think most of the trouble with the borrow checker could be solved by using free functions instead of a struct.

Reply:
Hmm, I'm not sure! I think the problem is that we iterate over `self:constraint_system.bus_interactions` while mutating `self`.

This is the error I get when I remove the collect:
```
error[E0502]: cannot borrow `*self` as mutable because it is also borrowed as immutable
   --> constraint-solver/src/solver.rs:102:29
    |
94  | /                 self.constraint_system
95  | |                     .bus_interactions
    | |_____________________________________- immutable borrow occurs here
...
101 |               for effect in effects {
    |                             ------- immutable borrow later used here
102 |                   progress |= self.apply_effect(effect);
    |                               ^^^^^^^^^^^^^^^^^^^^^^^^^ mutable borrow occurs here
```